src/category_theory/limits/shapes/pullbacks.lean

```diff
@@ -0,0 +1,244 @@
+@[simp] lemma span_map_snd {X Y Z : C} (f : X ⟶ Y) (g : X ⟶ Z) :
+  (span f g).map walking_span_hom.snd = g := rfl
+
+@[simp] lemma cospan_map_id {X Y Z : C} (f : X ⟶ Z) (g : Y ⟶ Z) (w : walking_cospan) :
+  (cospan f g).map (walking_cospan_hom.id w) = 𝟙 _ := rfl
+@[simp] lemma span_map_id {X Y Z : C} (f : X ⟶ Y) (g : X ⟶ Z) (w : walking_span) :
+  (span f g).map (walking_span_hom.id w) = 𝟙 _ := rfl
+
+
+variables {X Y Z : C}
+
+attribute [simp] walking_cospan_hom_id walking_span_hom_id
+
+def pullback_cone (f : X ⟶ Z) (g : Y ⟶ Z) := cone (cospan f g)
+
+namespace pullback_cone
+variables {f : X ⟶ Z} {g : Y ⟶ Z}
+
+def π₁ (t : pullback_cone f g) : t.X ⟶ X := t.π.app left
+def π₂ (t : pullback_cone f g) : t.X ⟶ Y := t.π.app right
+
+def mk {W : C} (π₁ : W ⟶ X) (π₂ : W ⟶ Y)
+  (eq : π₁ ≫ f = π₂ ≫ g) :
+  pullback_cone f g :=
+{ X := W,
+  π :=
+  { app := λ j, walking_cospan.cases_on j π₁ π₂ (π₁ ≫ f),
+    naturality' := λ j j' f, begin cases f; obviously end } }
+
+def condition (t : pullback_cone f g) : (π₁ t) ≫ f = (π₂ t) ≫ g :=
+begin
+  erw [t.w inl, ← t.w inr], refl
+end
+
+end pullback_cone
+
+def pushout_cocone (f : X ⟶ Y) (g : X ⟶ Z) := cocone (span f g)
+
+namespace pushout_cocone
+
+variables {f : X ⟶ Y} {g : X ⟶ Z}
+
+def ι₁ (t : pushout_cocone f g) : Y ⟶ t.X := t.ι.app left
+def ι₂ (t : pushout_cocone f g) : Z ⟶ t.X := t.ι.app right
+
+def mk {W : C} (ι₁ : Y ⟶ W) (ι₂ : Z ⟶ W)
+  (eq : f ≫ ι₁ = g ≫ ι₂) :
+  pushout_cocone f g :=
+{ X := W,
+  ι :=
+  { app := λ j, walking_span.cases_on j (f ≫ ι₁) ι₁ ι₂,
+    naturality' := λ j j' f, by cases f; obviously } }
+
+def condition (t : pushout_cocone f g) : f ≫ (ι₁ t) = g ≫ (ι₂ t) :=
+begin
+  erw [t.w fst, ← t.w snd], refl
+end
+
+end pushout_cocone
+
+def cone.of_pullback_cone
+  {F : walking_cospan.{v} ⥤ C} (t : pullback_cone (F.map inl) (F.map inr)) : cone F :=
+{ X := t.X,
+  π :=
+  { app := λ X, t.π.app X ≫ eq_to_hom (by tidy),
+    naturality' := λ j j' g,
+    begin
+      cases j; cases j'; cases g; dsimp; simp,
+      erw ← t.w inl, refl,
+      erw ← t.w inr, refl,
+    end } }.
+
+@[simp] lemma cone.of_pullback_cone_π
+  {F : walking_cospan.{v} ⥤ C} (t : pullback_cone (F.map inl) (F.map inr)) (j):
+  (cone.of_pullback_cone t).π.app j = t.π.app j ≫ eq_to_hom (by tidy) := rfl
+
+def cocone.of_pushout_cocone
+  {F : walking_span.{v} ⥤ C} (t : pushout_cocone (F.map fst) (F.map snd)) : cocone F :=
+{ X := t.X,
+  ι :=
+  { app := λ X, eq_to_hom (by tidy) ≫ t.ι.app X,
+    naturality' := λ j j' g,
+    begin
+      cases j; cases j'; cases g; dsimp; simp,
+      erw ← t.w fst, refl,
+      erw ← t.w snd, refl,
+    end } }.
+
+@[simp] lemma cocone.of_pushout_cocone_ι
+  {F : walking_span.{v} ⥤ C} (t : pushout_cocone (F.map fst) (F.map snd)) (j):
+  (cocone.of_pushout_cocone t).ι.app j = eq_to_hom (by tidy) ≫ t.ι.app j := rfl
+
+def pullback_cone.of_cone
+  {F : walking_cospan.{v} ⥤ C} (t : cone F) : pullback_cone (F.map inl) (F.map inr) :=
+{ X := t.X,
+  π :=
```

Comment:
I would put the `{ app ... }` on this line, since it's a one-liner.

Reply:
Done